File: leetcode/top-interview-150/169/Java/Solution.java
import java.util.HashMap;
import java.util.Map;

import com.apple.laf.resources.aqua;

public class Solution {

    public static void main(String[] args) {

    }

    public static int majorityElement(int[] nums) {

        HashMap<Integer, Integer> map = new HashMap<>();

        int n = nums.length;
        for (int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }

        for (Map.Entry<Integer, Integer> entry : map.entrySet()){
            if(entry.getValue() > n/2){
                return entry.getKey();
            }
        }

        return 0;
    }

    public static int theBoyerMooreVoting ( int[] nums){
        int candidate = 0;
        int n = nums.length;
        int count = 0;
        for(int i = 0; i < n; i++){
            if( count = 0){
                candidate = i;
            }
            if( nums[i] = nums[candidate]){
                count++;
            } else {
                count--;
            }
        }
        return nums[candidate];
    }

}

File: leetcode/top-interview-150/169/Solution.go
package main

import "fmt"

func main() {
	// nums := []int{2, 2, 1, 1, 1, 2, 2}
	nums := []int{6, 5, 5}
	fmt.Print(majorityElement(nums))
	fmt.Print(boyerMooreVotingAlgorithm(nums))
}

func majorityElement(nums []int) int {
	freq := make(map[int]int)

	for _, num := range nums {
		freq[num]++
	}

	maxFreq := 0
	resutl := 0

	for key, value := range freq {
		if value > maxFreq {
			maxFreq = value
			resutl = key
		}
	}

	return resutl
}

func boyerMooreVotingAlgorithm(nums []int) int {
	candidate := 0
	count := 0
	for _, num := range nums {
		if count == 0 {
			candidate = num
			count++
		} else if num == candidate {
			count++
		} else {
			count--
		}
	}
	return candidate
}
